include/zephyr/bluetooth/encrypted_ad_data.h

```diff
@@ -0,0 +1,77 @@
+/* Copyright (c) 2023 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <zephyr/kernel.h>
+
+#define BT_EAD_AAD_SIZE	       1  /**< Additional Authenticated Data size in bytes */
+#define BT_EAD_RANDOMIZER_SIZE 5  /**< Randomizer size in bytes */
+#define BT_EAD_NONCE_SIZE      13 /**< nonce size in bytes */
+#define BT_EAD_KEY_SIZE	       16 /**< Key size in bytes */
+#define BT_EAD_IV_SIZE	       8  /**< Initialisation Vector size in bytes */
+#define BT_EAD_MIC_SIZE	       4  /**< MIC size in bytes */
+
+/* This value is used to set the directionBit of the CCM  nonce to the MSB of the Randomizer field
+ * (see Supplement to the Bluetooth Core   Specification v11, Part A 1.23.3)
+ */
+#define BT_EAD_RANDOMIZER_DIRECTION_BIT 7
+
+/* Fixed value used for the Additional Authenticated Data (see Supplement to the Bluetooth Core
+ * Specification v11, Part A 1.23.3)
+ */
+static uint8_t bt_ead_aad[] = {0xEA};
+BUILD_ASSERT(sizeof(bt_ead_aad) == BT_EAD_AAD_SIZE);
+
+/**
+ * @brief Encrypt and authenticate the given advertising data.
+ * The resulting data in \p encrypted_ad_data will look like that:
+ *     - Randomizer is added in the 5 first bytes;
+ *     - AAD is added in the next byte;
+ *     - Encrypted payload is added (payload_size bytes);
+ *     - MIC is added in the last 4 bytes.
+ *
+ * @note The function must be called each time the RPA is updated or the data
+ * are modified.
+ *
+ * @note The term `advertising strucure` is used to describe the advertising
+ *       data with the advertising type and the length of those two.
+ *
+ * @param[in]  session_key Key of 16 bytes used for the encryption.
+ * @param[in]  iv Initialisation Vector used to generate the nonce. It must be
+ *             changed each time the Session Key change.
+ * @param[in]  payload Advertising Data to encrypt. Can be multiple advertising
+ *             structure concatenated.
+ * @param[in]  payload_size Size of the Advertising Data to encrypt.
+ * @param[out] encrypted_ad_data Encrypted Adv Data including the Randomizer and
+ *             the MIC. Size must be `BT_EAD_RANDOMIZER_SIZE + BT_EAD_AAD_SIZE +
+ *             payload_size + BT_EAD_MIC_SIZE`.
+ * @return int Size of \p encrypted_ad_data in bytes if successful, else return
+ *             `-EIO` if an error occurred during the encryption.
+ */
+int bt_ead_encrypt_ad_data(const uint8_t *session_key, const uint8_t *iv, const uint8_t *payload,
+			   size_t payload_size, uint8_t *encrypted_ad_data);
```

Comment:
That's a good point. But on the other hand, that's exactly why the host is so bloated. A lot of tiny features added to it because it 'makes sense' and we end up with a mess that's pretty hard to reason about and debug.

Reply:
Also, having it as an option as you describe cancel the option for the user to have both encrypted and non-encrypted ad data.

Reply:
> Also, having it as an option as you describe cancel the option for the user to have both encrypted and non-encrypted ad data.

I am not fully aware of the details of the feature: Are you encrypting the advertising data, or just a single LTV structure in it?

Reply:
> That's a good point. But on the other hand, that's exactly why the host is so bloated. A lot of tiny features added to it because it 'makes sense' and we end up with a mess that's pretty hard to reason about and debug.

I don't disagree, however from a user perspective it's much easier to use. It's basically a question of UX version how easy it is to maintain, right?

Reply:
That's subjective I think. What you find easier to use and what I (or someone else) finds easier to use can be two very different things. Keep in mind that you and I have been working with this API for a while, and that it may cloud our judgment a bunch.

What I've seen is that in practice this nice UX promise ends up not being true: see the myriad of ways to call the GATT API - added for 'nicer UX' - ending up with footguns where the structs you pass are modified depending on what's in them, the stack behaving subtly differently when seemingly-unrelated kconfig options are enabled, stuff like that.

The advertising module is already complicated enough to use and maintain, as have shown some recent bugs.

My point of view, and I think @alwa-nordic and @theob-pro also agree, is that we should have a small and opinionated host API, that is used by pluggable host/lib modules. This way, the complexity and internal state re-use is minimized and the users can have a nice-to-use API with less bugs. 
The downside of course is the impact on code size and speed, but IMO it's better to have slower code that is correct rather than something that is fast but really buggy.

tldr: no, it's usually not that easier to use in practice and yes it makes it much harder to maintain.

Reply:
> I am not fully aware of the details of the feature: Are you encrypting the advertising data, or just a single LTV structure in it?

Only the Ad Data are encrypted, and placed in a new Ad Type `Encrypted Advertising Data`. The encrypted data can contain multiple advertising structure. The advertising payload can contain an encrypted advertising structure and a local name structure. The Ad Data of the encrypted advertising structure will be encrypted but not the one of the local name structure.

Reply:
> tldr: no, it's usually not that easier to use in practice and yes it makes it much harder to maintain.

Indeed.

After @theob-pro 's helpful comment, I am fine with the current implementation. I had thought that the entire advertising data was encrypted, and not the individual LTV structures. So my suggestion is no longer valid, as that provides not only a less flexible solution, but also that it is easier to decode the data than I thought, as you simply just check the AD type after using `bt_data_parse` to decrypt the individual LTVs that may be encrypted. 

TLDR: I am fine with the suggested API